http-client/Network/HTTP/Client/Request.hs

```diff
@@ -77,24 +80,33 @@ parseUrl :: MonadThrow m => String -> m Request
 parseUrl = parseUrlThrow
 {-# DEPRECATED parseUrl "Please use parseUrlThrow, parseRequest, or parseRequest_ instead" #-}
 
--- | Same as 'parseRequest', except will throw an 'HttpException' in
--- the event of a non-2XX response.
+-- | Same as 'parseRequest', except will throw an 'HttpException' in the
+-- event of a non-2XX response. This uses 'throwErrorStatusCodes' to
+-- implement 'checkResponse'.
 --
 -- @since 0.4.30
 parseUrlThrow :: MonadThrow m => String -> m Request
 parseUrlThrow =
     liftM yesThrow . parseRequest
   where
-    yesThrow req = req
-        { checkResponse = \_req res ->
-            let W.Status sci _ = responseStatus res in
-            if 200 <= sci && sci < 300
-                then return ()
-                else do
-                    chunk <- brReadSome (responseBody res) 1024
-                    let res' = fmap (const ()) res
-                    throwHttp $ StatusCodeException res' (L.toStrict chunk)
-        }
+    yesThrow req = req { checkResponse = \_req -> throwErrorStatusCodes }
+
+-- | Throws a 'StatusCodeException' if the response's status code
+-- indicates an error (if it isn't 2xx).
+--
+-- @const throwErrorStatusCodes@ can be used to implement
+-- 'checkResponse'.
+--
+-- @since 0.5.12.2
+throwErrorStatusCodes :: MonadIO m => Response BodyReader -> m ()
+throwErrorStatusCodes res = do
+    let W.Status sci _ = responseStatus res
+    if 200 <= sci && sci < 300
+        then return ()
+        else liftIO $ do
+            chunk <- brReadSome (responseBody res) 1024
+            let res' = fmap (const ()) res
+            throwHttp $ StatusCodeException res' (L.toStrict chunk)
```

Comment:
We should not export this function, due to the unwrapped call to `throwHttp`. `throwHttp` uses the hidden data type `HttpExceptionContentWrapper`, which is always unwrapped by this library. If we expose this function, there are no guarantees that will happen.

Reply:
I see, good catch!  I've modified this function to take `Request` as an argument and directly throw a `HttpExceptionRequest`.